diffsmhm/diff_stats/cuda/wprp.py

```diff
@@ -432,37 +444,78 @@ def wprp_mpi_kernel_cuda(
         A named tuple of the data needed for the MPI reduction and final summary stats.
     """
-    assert not np.allclose(rpbins_squared[0], 0)
-    _rpbins_squared = np.concatenate([[0], rpbins_squared], axis=0)
+    assert not cp.allclose(rpbins_squared[0], 0)
+    _rpbins_squared = cp.concatenate([cp.array([0]), rpbins_squared], axis=0)
 
     n_grads = w1_jac.shape[0]
     n_rp = _rpbins_squared.shape[0] - 1
     n_pi = int(zmax)
 
-    result = cuda.to_device(np.zeros(n_rp * n_pi, dtype=np.float64))
-    result_grad = cuda.to_device(
-        np.zeros(n_grads * n_rp * n_pi, dtype=np.float64)
-    )
-    _count_weighted_pairs_rppi_with_derivs_cuda[blocks, threads](
-        x1, y1, z1, w1, w1_jac, inside_subvol,
-        _rpbins_squared, n_pi,
-        result, result_grad,
-    )
-    res = result.copy_to_host().reshape((n_rp, n_pi))
-    res_grad = result_grad.copy_to_host().reshape((n_grads, n_rp, n_pi))
+    # EDIT: call onto multiple GPUs if available
+    n_devices = len(cuda.gpus)
+
+    # split the data into chunks for each device
+    avg, rem = divmod(len(x1), n_devices)
+    device_count = [avg + 1 if p < rem else avg for p in range(n_devices)]
+    device_displ = [sum(device_count[:p]) for p in range(n_devices)]
+    # lists of results to combine later
+    result_all = []
+    result_grad_all = []
+    for d in range(n_devices):
+        # get this chunk range
+        count = device_count[d]
+        displ = device_displ[d]
+        start_idx = displ
+        end_idx = displ + count
+
+        # copy data to relevant device
+        cp.cuda.Device(d).use()
+        x1_d = cp.copy(x1)
+        y1_d = cp.copy(y1)
+        z1_d = cp.copy(z1)
+
+        w1_d = cp.copy(w1)
+        w1_jac_d = cp.copy(w1_jac)
+
+        _rpbins_squared_d = cp.copy(_rpbins_squared)
+        inside_subvol_d = cp.copy(inside_subvol)
+
+        result_d = cp.zeros(n_rp * n_pi, dtype=np.float64)
+        result_grad_d = cp.zeros(n_grads * n_rp * n_pi, dtype=np.float64)
+
+        _count_weighted_pairs_rppi_with_derivs_cuda[blocks, threads](
+            x1_d, y1_d, z1_d, w1_d, w1_jac_d, inside_subvol_d,
+            _rpbins_squared_d, n_pi,
+            result_d, result_grad_d,
+            start_idx, end_idx
+        )
 
-    # TODO - do this with cupy if it is available?
-    sums = cuda.to_device(np.zeros(2 + 2*n_grads, dtype=np.float64))
-    _sum_mask[blocks, threads](w1, inside_subvol, sums, 0)
-    _sum2_mask[blocks, threads](w1, inside_subvol, sums, 1)
+        result_all.append(result_d)
+        result_grad_all.append(result_grad_d)
+
+    # now add the distributed calculation
+    res = cp.zeros(n_rp * n_pi, dtype=np.float64)
+    res_grad = cp.zeros(n_grads * n_rp * n_pi, dtype=np.float64)
+    for d in range(n_devices):
+        res += result_all[d]
+
+        res_grad += result_grad_all[d]
+
+    res = cp.reshape(res, (n_rp, n_pi))
+    res_grad = cp.reshape(res_grad, (n_grads, n_rp, n_pi))
```

Comment:
Does cupy have any docs / hints on how it combines this data across devices? In other words, is it going to do the "right" thing often enough that we are ok with this, or do we want to have a more explicit approach here?

Reply:
There's a performance warning that it is transferring data across devices, and the results match the serial cpu kernel so it seems to be doing the addition correctly. But I do want to have a more explicit approach in the future to avoid the warning message or figure out the fastest way to do that transfer